relay-server/src/services/processor/replay.rs

```diff
@@ -113,6 +130,12 @@ pub fn process(
         }
     });
 
+    if dropped_item {
+        state
+            .managed_envelope
+            .reject(Outcome::Invalid(DiscardReason::ReplayRejected));
+    }
```

Comment:
This leads to different behavior depending on the order of items:

1. If the first item is dropped with an outcome, all subsequent items are dropped silently (one outcome in total).
2. If the second item is dropped with an outcome, the first outcome will still be in the envelope after `retain_items`. In that case, the first item will be rejected with `ReplayRejected` outcome (two outcomes in total).

Maybe we can replace `retain_items` with a simple for loop that breaks on the first failure, and then rejects the envelope with the outcome of the first failure.

Reply:
`reject` does log outcomes (the `Handling` is only used to decide whether or not to log an error). But it turns out we don't log outcomes for replays in `reject`, which is probably just oversight: https://github.com/getsentry/relay/blob/8eff86344e5e313583c131892f29bb825178d7da/relay-server/src/utils/managed_envelope.rs#L433-L468

Not sure what the best way forward is here, maybe the best would be to return a `ProcessingError`, which is then automatically reported as an outcome via `ProcessingError::to_outcome`.

Reply:
I've added `retain_or_reject_all` methods to `ManagedEnvelope` and `Envelope`.  It guarantees one outcome is emitted, breaks early on processing error, ensures the envelope is empty after error, and uses all of the familiar APIs replay is accustomed to.

Reply:
I looked into the code some more, I'm convinced now that the cleanest way to reject an entire envelope is to raise a `ProcessingError` and let it bubble up, because that's API that already exists for this purpose. And to make sure that `ManagedEnvelope::reject` emits outcomes for replays.

For the test, you can add the `outcomes_consumer` fixture in existing replay integration tests for that, or add new tests in `test_outcome.py`.

Reply:
@jjbayer This commit implements the ProcessingError flow https://github.com/getsentry/relay/pull/3201/commits/27fcc83e9fa23cbd23b3435f570e80fb3db94aaa